File: solutions/csharp/bob/2/Bob.cs
using System.Linq;

public static class Bob
{
    public static string Response(string statement) =>
        statement.Trim() switch
        {
            string s when s.EndsWith('?') && IsYelling(s) => "Calm down, I know what I'm doing!",
            string s when s.EndsWith('?') => "Sure.",
            string s when IsYelling(s) => "Whoa, chill out!",
            string s when s == string.Empty => "Fine. Be that way!",
            _ => "Whatever."
        };

    private static bool IsYelling(string s)
    {
        return s.ToUpper() == s && s.Any(char.IsLetter);
    }
}

File: solutions/fsharp/bob/2/Bob.fs
module Bob

let response (input: string) : string =
    let (|Question|_|) (phrase: string) =
        if phrase.EndsWith '?' then Some() else None

    let (|Yell|_|) (phrase: string) =
        if Seq.exists System.Char.IsLetter phrase && phrase.ToUpper() = phrase then
            Some()
        else
            None

    let (|Silence|_|) (phrase: string) = if phrase = "" then Some() else None

    match input.Trim() with
    | Yell & Question -> "Calm down, I know what I'm doing!"
    | Yell -> "Whoa, chill out!"
    | Question -> "Sure."
    | Silence -> "Fine. Be that way!"
    | _ -> "Whatever."
